parsons/action_builder/action_builder.py

```diff
@@ -0,0 +1,351 @@
+        """  # noqa: E501
+
+        # Check that we have appropriate entity type/identifier, name, and campaign first
+        if entity_type == identifiers is None:
+            error_msg = "Must provide either entity_type (to insert a new record) "
+            error_msg += "or identifiers (to update an existing record)"
+            raise ValueError(error_msg)
+
+        if not isinstance(data, dict):
+            data = {}
+
+        name_check = [
+            key for key in data.get("person", {}) if key in ("name", "given_name")
+        ]
+        if identifiers is None and not name_check:
+            raise ValueError("Must provide name or given name if inserting new record")
+
+        campaign = self._campaign_check(campaign)
+
+        url = f"campaigns/{campaign}/people"
+
+        if "person" not in data:
+            # The POST data must live inside of person key
+            data["person"] = {}
+
+        if identifiers:
+            # Updating an existing record
+
+            if isinstance(identifiers, str):
+                # Ensure identifiers is a list, even if it will usually just be one item
+                identifiers = [identifiers]
+
+            # Default to assuming identifier comes from Action Builder and add prefix if missing
+            identifiers = [
+                f"action_builder:{x}" if ":" not in x else x for x in identifiers
+            ]
+            data["person"]["identifiers"] = identifiers
+
+        if entity_type:
+            # Inserting a new record
+            data["person"]["action_builder:entity_type"] = entity_type
+
+        return self.api.post_request(url=url, data=json.dumps(data))
+
+    def add_tags_to_record(
+        self, identifiers, tag_name, tag_field, tag_section, campaign=None
+    ):
+        """
+        Add a tag (i.e. custom field value) to an existing entity record in Action Builder. The
+        tags, along with their category/field and section/group, must already exist (except for
+        date fields). If fewer fields are provided than tags, the excess tags will be applied to
+        the final field provided. Similarly, the final section will be applied to any fields or
+        tags in excess.
+        `Args:`
+            identifiers: list
+                The list of strings of unique identifiers for a record being updated. ID strings
+                will need to begin with the origin system, followed by a colon, e.g.
+                `action_builder:abc123-...`.
+            tag_name: str or list
+                The name(s) of the new tag(s), i.e. the custom field value(s).
+            tag_field: str or list
+                The name(s) of the tag category(ies), i.e. the custom field name(s). Must have
+                equal or fewer items as `tag_name`.
+            tag_section: str or list
+                The name(s) of the tag section(s), i.e. the custom field group name(s). Must have
+                equal or fewer items as `tag_section`.
```

Comment:
I may not be understanding the data structure here, but I wonder if it would be easier to accept a list of dictionaries:

[
   { tag_name: 'foo', tag_fields: ['bar', 'baz'], tag_sections: ['candy', 'cane'] },
   ...
]

Reply:
seems like that might make the parsing trivial and would force the user to be explicit about what relationships they wanted between names, fields, sections

Reply:
I'm willing to be swayed here, but I wanted to name a bit more of the intent first. This is meant as a convenience method, because someone who is more comfortable with the JSON needed to apply tags could just include that in the `data` provided to `_upsert_entity()`. It's a list of dicts to be added under a key called `add_tags` at the same level as `person`.

So my reasoning is that someone would only use this method if they want a more intuitive approach to adding tags than constructing that JSON. It's also annoying because every tag submitted must have a field and a section in this JSON data as well, and I wouldn't be surprised if it's often the case that all tags applied this way are in the same section.

The current logic I have is that if the lists provided are not the same length, the last field/section continues to be applied for any remaining tags. That way, you could just provide one section, and it would apply to all the tags. If that's _not_ intuitive, then I definitely haven't achieved my goal here, but I'm worried that asking for a list of dicts would essentially make this method no longer any more convenient than expecting that to be added to the `data` in any of the upsert family of methods...

Reply:
Interesting. I see what you're going for but I definitely did not intuit the last value continues to be applied to the remaining tags.

Reply:
Then I should definitely see what I can do about making that clearer in the comments/docstrings!

Reply:
Ok, just pushed some updates to comments and docstrings, in case they help